File: DebugIntro/Fortran/01_RoundingAndTruncation.f90

MODULE KINDS

  !Constants for "single" and "double" precision floats
  INTEGER, PARAMETER :: real32 = KIND(1.0)
  INTEGER, PARAMETER :: real64 = KIND(1.0d0)

END MODULE KINDS


PROGRAM example

  USE KINDS

  IMPLICIT NONE

  INTEGER :: i, j, n
  REAL(KIND=real32) :: a = 3.0, b = 5.0, c
  REAL(KIND=real64) :: d = 3.14159265358979323846_real64
  !Note the _real64 wart which makes this a 64-bit literal
  CHARACTER(22) :: str = "3.14159265358979323846"
  REAL(KIND=real64) :: f, g


  i = a / b
  c = d

  PRINT*, "Floats are truncated to store into integers:"
  WRITE(*,fmt='(f6.3, a2, f6.3, a18, i0)')  a, ' /', b, "as an integer is ", i
  PRINT*, ''
  b = 2.0; i = a / b
  WRITE(*,fmt='(f6.3, a2, f6.3, a18, i0)')  a, ' /', b, "as an integer is ", i
  PRINT*, ''

  PRINT*, "All floats have limited precision:"
  WRITE(*, fmt='(a23, a15, f23.20, a23, f15.12)') str, &
      "as a double is", d, "and that as a float is", c

  f = 0.5; g = 0.01
  i = f/g
  WRITE(*, fmt='(f15.12, a2, f15.12, a29, i0)') f, '/', g, &
      "might not be 50! Here it is ", i

  PRINT*, ''
  PRINT*, &
      "Summing small numbers can give unexpected results due to finite precision:"

  f = 0.0
  n = 10000
  g = 1.0/n

  DO i = 1, n
    f = f + g
  ENDDO

  j = f
  WRITE(*, fmt='(f15.12, a8, i0, a10, f23.20, a15, i0)') g, "summed ", n, &
      "times is ", f, "and rounds to ", j

  j = 0; f = 0.0

  DO WHILE(f < 1.0)
    j = j + 1
    f = f + g
  ENDDO

  WRITE(*, fmt='(a7, f15.12, a34, i0, a28, i0)') "Using ", g, &
      "as a while loop increment we get ", j, " iterations where we expect ", n


END PROGRAM example

File: DebugIntro/C/01_RoundingAndTruncation.c
#include <stdio.h>
#include <float.h>


int main(int argc, char** argv){

  int i, j, n;
  float a = 3.0, b = 5.0, c;
  double d = 3.14159265358979323846;
  char *str = "3.14159265358979323846";
  double f, g;

  i = a/b;
  c = d;

  printf("Floats are truncated to store into integers:\n");
  printf("%f/%f as an integer is %d \n", a, b, i);

  b = 2.0; i = a/b;
  printf("%f/%f as an integer is %d \n", a, b, i);

  printf("\n");
  printf("All floats have limited precision:\n");

  printf("%s as a double is %23.20f and that as a float is %15.12f \n", str, d, c);

  f = 0.5; g=0.01;
  i = f/g;
  printf("%15.12f/%15.12f might not be 50! Here it is %d \n", f, g, i);

  printf("\nSumming small numbers can give unexpected results due to finite precision:\n");

  f = 0.0;
  n = 10000;
  g = 1.0/n;
  for(i=0; i < n; i++){
    f += g;
  }
  j = f;
  printf("%15.12f summed %d times is %23.20f and rounds to %d\n", g, n, f, j);

  j = 0; f = 0.0;
  while(f < 1.0){
    j++;
    f += g;
  }
  printf("Using %15.12f as a while loop increment we get %d iterations where we expect %d", g, j, n);


  printf("\n");

  return 0;

}

